boot/freeldr/freeldr/arch/i386/multiboot.S

```diff
@@ -98,12 +111,34 @@ mb3:
     /* Relocate itself to lower address */
     mov esi, INITIAL_BASE
     mov edi, FREELDR_BASE
+#ifndef _USE_ML
     mov ecx, offset __bss_start__ - FREELDR_BASE
+#else
+    /* Address the image with es segment */
+    // mov ax, FREELDR_PE_BASE / 16 /* == &__ImageBase */
+    // mov es, ax
+    mov eax, offset ___ImageBase
+    shr eax, 4
+    mov es, eax
+
+    /* Get address of optional header */
+    mov eax, dword ptr es:[IMAGE_DOS_HEADER_e_lfanew]
+    add eax, 4 + IMAGE_FILE_HEADER_SIZE
+
+    /* Get the size of the image */
+    mov ecx, dword ptr es:[eax + IMAGE_OPTIONAL_HEADER_SizeOfImage]
+    // add ecx, FREELDR_PE_BASE - FREELDR_BASE /* == offset __bss_start__ - FREELDR_BASE since offset __bss_start__ == &__ImageBase (== FREELDR_PE_BASE) + OptionalHeader->SizeOfImage */
+    add ecx, offset ___ImageBase - FREELDR_BASE
+#endif
```

Comment:
The idea of that block is that, since we don't know anything about BSS segment in the MSVC build I use some runtime way to guess the full image size of freeldr (including the 16-bit frldr16 stub that is prepended), based on the following code:
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/lib/mm/meminit.c;hb=d27286655018c15688afe2dd88f1ac79dbc72aa9#l310

The code in comment is what I want to achieve.

Also I noticed that `__ImageBase` is actually equal to FREELDR_PE_BASE so I used it instead.

Reply:
Any suggestions to improve that are welcome.

Reply:
Let's just merge it and get rid of this stuff entirely.